miner/block.go

```diff
@@ -0,0 +1,397 @@
+		txFeeRecipient = w.validator
+		log.Warn("TxFeeRecipient and Validator flags set before split etherbase fork is active. Defaulting to the given validator address for the coinbase.")
+	}
+
+	// Only set the coinbase if our consensus engine is running (avoid spurious block rewards)
+	if w.isRunning() {
+		if txFeeRecipient == (common.Address{}) {
+			return nil, errors.New("Refusing to mine without etherbase")
+		}
+		header.Coinbase = txFeeRecipient
+	}
+	// Note: The parent seal will not be set when not validating
+	if err := w.engine.Prepare(w.chain, header); err != nil {
+		log.Error("Failed to prepare header for mining", "err", err)
+		return nil, fmt.Errorf("Failed to prepare header for mining: %w", err)
+	}
+
+	// Initialize the block state itself
+	state, err := w.chain.StateAt(parent.Root())
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get the parent state: %w:", err)
+	}
+
+	b := &blockState{
+		signer:         types.NewEIP155Signer(w.chainConfig.ChainID),
+		state:          state,
+		tcount:         0,
+		gasLimit:       core.CalcGasLimit(parent, state),
+		header:         header,
+		txFeeRecipient: txFeeRecipient,
+	}
+	b.gasPool = new(core.GasPool).AddGas(b.gasLimit)
+
+	// Play our part in generating the random beacon.
+	if w.isRunning() && random.IsRunning() {
+		istanbul, ok := w.engine.(consensus.Istanbul)
+		if !ok {
+			log.Crit("Istanbul consensus engine must be in use for the randomness beacon")
+		}
+
+		lastCommitment, err := random.GetLastCommitment(w.validator, b.header, b.state)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to get last commitment: %w", err)
+		}
+
+		lastRandomness := common.Hash{}
+		if (lastCommitment != common.Hash{}) {
+			lastRandomnessParentHash := rawdb.ReadRandomCommitmentCache(w.db, lastCommitment)
+			if (lastRandomnessParentHash == common.Hash{}) {
+				return nil, errors.New("Failed to get last randomness cache entry")
+			}
+
+			var err error
+			lastRandomness, _, err = istanbul.GenerateRandomness(lastRandomnessParentHash)
+			if err != nil {
+				return nil, fmt.Errorf("Failed to generate last randomness: %w", err)
+			}
+		}
+
+		_, newCommitment, err := istanbul.GenerateRandomness(b.header.ParentHash)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to generate new randomness: %w", err)
+		}
+
+		err = random.RevealAndCommit(lastRandomness, newCommitment, w.validator, b.header, b.state)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to reveal and commit randomness: %w", err)
+		}
+		// always true (EIP158)
+		b.state.IntermediateRoot(true)
+
+		b.randomness = &types.Randomness{Revealed: lastRandomness, Committed: newCommitment}
+	} else {
+		b.randomness = &types.EmptyRandomness
+	}
+
+	return b, nil
+}
+
+// selectAndApplyTransactions selects and applies transactions to the in flight block state.
+func (b *blockState) selectAndApplyTransactions(ctx context.Context, w *worker) error {
```

Comment:
I don't like 2 as a solution (we're mutating blockState, but not the worker)
For 1: It would be possible, but a fair amount of work to pull disentangle the worker (this code relies on accessing a lot of data directly)